### PR TITLE
Update main.go

### PR DESCRIPTION
**Pull request by jjoshua2**
matches asynchrounous like my other pull, but also exponential retry on uploadGame. match still needs it if this is acceptable we can copy the code over

### CODE REVIEW

**Comment by jjoshua2** [comment]
Initially I used retriesLeft and decremented, but exponential backoff is superior so if internet is out overnight/weekend it will resume uploads a long time later, instead of every 5 seconds for a few tries.

File: go/src/client/main.go [hunk 1]
```diff
@@ -96,6 +96,8 @@ func uploadGame(httpClient *http.Client, path string, pgn string, nextGame clien
 	body := &bytes.Buffer{}
 	_, err = body.ReadFrom(resp.Body)
 	if err != nil {
+		time.Sleep(time.Millisecond * 2000 << retryCount)
+		uploadGame(httpClient, path, pgn, nextGame, retryCount+1)
```
Comment (glinscott) [inline go/src/client/main.go]: `err = uploadGame(...)`

File: go/src/client/main.go [hunk 2]
```diff
@@ -96,6 +96,8 @@ func uploadGame(httpClient *http.Client, path string, pgn string, nextGame clien
 	body := &bytes.Buffer{}
 	_, err = body.ReadFrom(resp.Body)
 	if err != nil {
+		time.Sleep(time.Millisecond * 2000 << retryCount)
```
Comment (glinscott) [inline go/src/client/main.go]: Precedence always a bit tricky with shifts, maybe `time.Second * (2 << retryCount)`?

File: go/src/client/main.go [hunk 3]
```diff
@@ -328,15 +330,15 @@ func nextGame(httpClient *http.Client) error {
...
 		result, pgn := playMatch(networkPath, candidatePath, params, nextGame.Flip)
-		client.UploadMatchResult(httpClient, *HOSTNAME, nextGame.MatchGameId, result, pgn, getExtraParams())
+		go client.UploadMatchResult(httpClient, *HOSTNAME, nextGame.MatchGameId, result, pgn, getExtraParams())
 		return nil
 	} else if nextGame.Type == "train" {
 		networkPath, err := getNetwork(httpClient, nextGame.Sha, true)
 		if err != nil {
 			return err
 		}
 		trainFile, pgn := train(networkPath, params)
```
Comment (glinscott) [inline go/src/client/main.go]: We could potentially even read in the trainFile into memory, and return it from `train`.  That way we don't have to worry about the race here.

File: go/src/client/main.go (hunk 2 above)
Comment (jjoshua2) [reply go/src/client/main.go]: I don't think order of operations matters here at all. Either way the shift is just a fast multiply by 2^power and multiplication is commutative.

**Comment by jjoshua2** [comment]
I decided to commit 2/3 of your changes. Not sure I know go well enough to do your trainFile suggestion.

File: go/src/client/main.go (hunk 3 above)
Comment (jjoshua2) [reply go/src/client/main.go]: I don't understand what the race is, since the path is different each time this is called

Comment (glinscott) [reply go/src/client/main.go]: The path is the same each time isn't it though?  `train-{pid}/training.0.gz` type of thing.  Process id remains the same (it's the process ID of the client process).

Comment (jjoshua2) [reply go/src/client/main.go]: @Error323 Can we use the random seed that was used an name each file training.+seed+.gz or will that cause problems with your training? Should fix the problem of duplicate paths and have the bonus that in a crash, we could upload result without losing it in memory.

Comment (jjoshua2) [reply go/src/client/main.go]: Actually training.0.gz is hardcoded in main.go with no access to seed, but I could use a new rnd call with little work.

**Comment by glinscott** [comment]
Yay!  Looks great :).

**Comment by jjoshua2** [comment]
I threw in the time elapsed as a bonus lol. I hope the filenames being different wont break Error323s training, but its already broken anyway :(

**Comment by glinscott** [comment]
The server ignores the filename, should be great, but I'm playing a test game just to be sure.

**Pull request merged**